Adds GDAS GEMPAK jobs to rocoto task mesh

HenryRWinterbottom opened the pull request:
# Description

This PR addresses issue #1219. Rocoto task `jobs/rocoto/gempak.sh` was updated to support `JGDAS_ATMOS_GEMPAK`. Rocoto task `jobs/rocoto/gempakmeta.sh` was added to support `JGFS_ATMOS_GEMPAK_META` and `JGDAS_ATMOS_GEMPAK_META_NCDC`.

  Resolves #1219 
  References #1222 #2073

**NOTE** This is a replicated PR open to resolve developer mishap with Github interface.

# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

The Rocoto XML configuration has been executed as follows.

```
user@host$: cd /scratch1/NCEPDEV/da/Henry.Winterbottom/trunk/global-workflow.develop.20231117/workflow
user@host$: ./setup_xml.py /scratch1/NCEPDEV/da/Henry.Winterbottom/work/GLOBAL_WORKFLOW/x002_gfsv17_issue_1219/x002_gfsv17_issue_1219.xml
```

The resulting Rocoto workflow is as follows.

```
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202103231200             gfsstage_ic                    51952235              QUEUED                   -         0           0.0
202103231200                 gfsfcst                           -                   -                   -         -             -
202103231200            gfspost_f000                    51952236              QUEUED                   -         0           0.0
202103231200            gfspost_f006                    51952237              QUEUED                   -         0           0.0
202103231200            gfspost_f012                    51952238              QUEUED                   -         0           0.0
202103231200                 gfsvrfy                           -                   -                   -         -             -
202103231200             gfsmetpg2g1                           -                   -                   -         -             -
202103231200             gfsmetpg2o1                           -                   -                   -         -             -
202103231200             gfsmetppcp1                           -                   -                   -         -             -
202103231200               gfsgempak                           -                   -                   -         -             -
202103231200           gfsgempakmeta                           -                   -                   -         -             -
202103231200       gfsgempakmetancdc                           -                   -                   -         -             -
202103231200    gfsgempakncdcupapgif                           -                   -                   -         -             -
202103231200      gfsgempakpgrb2spec                           -                   -                   -         -             -
202103231200                 gfsarch                           -                   -                   -         -             -
202103231200              gfscleanup                           -                   -                   -         -             -

```

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

aerorahul commented on workflow/applications/gfs_forecast_only.py:
```diff
@@ -117,6 +117,9 @@ def get_task_names(self):
 
         if self.do_gempak:
             tasks += ['gempak']
+            tasks += ['gempakmeta']
+            tasks += ['gempakncdcupapgif']
+            tasks += ['gempakpgrb2spec']
```
This is fine now because we are considering `forecast-only` to be used only with `RUN=gfs`.  In the past, one could have used `forecast-only` with `RUN=gdas`.  It will take some effort to reinstate that capability.
This is just a note; no action required.

aerorahul commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -894,15 +893,46 @@ def gempak(self):
+        deps.append(rocoto.add_dependency(dep_dict))
+        dependencies = rocoto.create_dependency(dep=deps)
+
+        resources = self.get_resource('gempak')
```
I think you forgot to create `task` with `create_wf_task` for this method.

aerorahul reviewed (changes requested):
Looks ok with the exception that the `gempakncdcupagif` task is not created in the rocoto mesh.  This should error out when running `setup_xml.py`; so wondering why it doesn't?

HenryRWinterbottom commented:
> Looks ok with the exception that the `gempakncdcupagif` task is not created in the rocoto mesh. This should error out when running `setup_xml.py`; so wondering why it doesn't?

Should this be addressed in a different issue?

KateFriedman-NOAA commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -894,15 +893,47 @@ def gempak(self):
 
         return task
 
-    def npoess(self):
```
No `npoess_pgrb2_0p5deg` job?
https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/ecf/defs/gfs_v16_3.def#L2190
https://github.com/NOAA-EMC/global-workflow/blob/develop/ecf/scripts/gfs/atmos/gempak/jgfs_atmos_npoess_pgrb2_0p5deg.ecf

HenryRWinterbottom replied:
@KateFriedman-NOAA Are able to confirm that a relative change was (mistakenly) committed?

If not, perhaps we open a separate issue which I can address immediately.

KateFriedman-NOAA replied:
Yeah, so looking at the current `gfs_tasks.py` in `develop` I see the `npoess` job:
https://github.com/NOAA-EMC/global-workflow/blob/develop/workflow/rocoto/gfs_tasks.py#L897:L907

Can you add this job back in, update its name to `npoess_pgrb2_0p5deg`, and add the other pieces needed to add it to the rocoto mesh? I added another comment about this job being a dependency for one of the others. Based on ops triggers: https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/ecf/defs/gfs_v16_3.def#L2193

Thanks!

KateFriedman-NOAA reviewed (commented):
Appear to be missing the `npoess_pgrb2_0p5deg` job but otherwise the changes to add the other gempak jobs into the rocoto mesh look good. See inline comment with information on that job, which is a task dependency for the `pgrb2_spec` job.

There will be more work to test and troubleshoot these jobs but that is out-of-scope for this PR.

KateFriedman-NOAA commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -894,15 +893,47 @@ def gempak(self):
+
+    def gempakpgrb2spec(self):
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{self.cdump}post'}
```
This should have a job dependency on the `npoess_pgrb2_0p5deg` job.

KateFriedman-NOAA replied:
This still needs to be updated. :)

HenryRWinterbottom commented:
@aerorahul @KateFriedman-NOAA @WalterKolczynski-NOAA 

Thank you for taking the time to review this PR. After all of my botches last week, I believe that this PR has updated all of the comments, issues, etc., that are noted above. Please let me know otherwise. If nothing seems to be incorrect, this PR should be read to review (again). Thank you for your patience.

KateFriedman-NOAA commented on workflow/applications/gfs_cycled.py:
```diff
@@ -232,13 +236,16 @@ def get_task_names(self):
 
         if self.do_npoess:
-            gfs_tasks += ['npoess']
+            gfs_tasks += ['npoess_pgrb2_0p5deg']
```
The `npoess_pgrb2_0p5deg` job is part of the gempak family of jobs so you can move this line:
```
gfs_tasks += ['npoess_pgrb2_0p5deg']
```
...up under the `if self.do_gempak:` block above and do away with `do_npoess` flag wherever it is set/used.

KateFriedman-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -22,10 +22,11 @@ class Tasks:
                    'waveawipsbulls', 'waveawipsgridded', 'wavegempak', 'waveinit',
                    'wavepostbndpnt', 'wavepostbndpntbll', 'wavepostpnt', 'wavepostsbs', 'waveprep',
-                   'npoess']
+                   'npoess_pgrb2_0p5deg']
```
This is a member of the gempak job family so you can group it with the other gempak jobs.

KateFriedman-NOAA reviewed (changes requested):
Most prior comments addressed, few more to address. See new comments. Thanks @HenryWinterbottom-NOAA !

WalterKolczynski-NOAA commented:
@HenryWinterbottom-NOAA Kate pointed out an incorrect dependency for one of the jobs that should be corrected, plus there is now a conflict from the changes in your AWIPS PR.

HenryRWinterbottom commented:
@WalterKolczynski-NOAA I am going to close this PR and open a new that addresses all of the comments. My branch has become broken and convoluted and I can no longer keep up-to-date with the `develop` branch and maintain all of my additions. I will keep this open such that I have the comments but close once I open the new PR. Sorry for the inconvenience this causes.
